Subject: [PATCH] schedule vsync immediately when requested from the Looper thread

This patch allows Choreographer to schedule vsync immediately when it
is requested from the Looper thread. Previously this is not allowed
when vsync is already scheduled from a binder thread and hence there
is a delay up to 16ms (one vsync interval) in processing frame
callbacks.

diff --git a/core/java/android/view/Choreographer.java b/core/java/android/view/Choreographer.java
index d3db74d..bd8b9ff 100644
--- a/core/java/android/view/Choreographer.java
+++ b/core/java/android/view/Choreographer.java
@@ -557,6 +557,19 @@
                 msg.setAsynchronous(true);
                 mHandler.sendMessageAtTime(msg, nextFrameTime);
             }
+        } else {
+            if (USE_VSYNC) {
+                if (mHandler.hasMessages(MSG_DO_SCHEDULE_VSYNC)) {
+                    // If running on the Looper thread, then schedule the vsync immediately.
+                    if (isRunningOnLooperThreadLocked()) {
+                        if (DEBUG_FRAMES) {
+                            Log.d(TAG, "Scheduling next frame on vsync.");
+                        }
+                        scheduleVsyncLocked();
+                        mHandler.removeMessages(MSG_DO_SCHEDULE_VSYNC);
+                    }
+                }
+            }
         }
     }
 
